pkg/agent/controller/networkpolicy/l7engine/reconciler.go

```diff
@@ -0,0 +1,398 @@
+	}
+	return nil
+}
+
+func (r *Reconciler) deleteSuricataTenant(vlanID uint32) error {
+	// Unregister the tenant handler.
+	scCmd := fmt.Sprintf("unregister-tenant-handler %d vlan %d", vlanID, vlanID)
+	if output, err := r.suricataScFn(scCmd); err != nil {
+		return fmt.Errorf("failed to unregister handler for tenant %d with vlan %d: %v", vlanID, vlanID, output)
+	}
+
+	// Unregister the tenant.
+	scCmd = fmt.Sprintf("unregister-tenant %d", vlanID)
+	if output, err := r.suricataScFn(scCmd); err != nil {
+		return fmt.Errorf("failed to unregister tenant %d: %v", vlanID, output)
+	}
+
+	// Delete the tenant config file.
+	configPath := generateTenantConfigPath(vlanID)
+	if err := defaultFS.Remove(configPath); err != nil {
+		return fmt.Errorf("failed to delete tenant config file %s: %v", configPath, err)
+	}
+	return nil
+}
+
+func (r *Reconciler) initializeSuricata() {
+	// Create the config file /etc/suricata/antrea.yaml for Antrea which will be included in the default Suricata config file
+	// /etc/suricata/suricata.yaml.
+	suricataAntreaConfigData := fmt.Sprintf(`%%YAML 1.1
+---
+af-packet:
+  - interface: %[1]s
+    threads: auto
+    cluster-id: 80
+    cluster-type: cluster_flow
+    defrag: no
+    use-mmap: yes
+    tpacket-v2: yes
+    checksum-checks: no
+    bpf-filter: "ip or ip6"
+    copy-mode: ips
+    copy-iface: %[2]s
+  - interface:  %[2]s
+    threads: auto
+    cluster-id: 81
+    cluster-type: cluster_flow
+    defrag: no
+    use-mmap: yes
+    tpacket-v2: yes
+    checksum-checks: no
+    bpf-filter: "ip or ip6"
+    copy-mode: ips
+    copy-iface: %[1]s
+multi-detect:
+  enabled: yes
+  selector: vlan
+`, config.L7NetworkPolicyTargetPortName, config.L7NetworkPolicyReturnPortName) // TODO: expand loaders when there is not enough loaders
+	f, err := defaultFS.Create(antreaSuricataConfigPath)
+	if err != nil {
+		klog.ErrorS(err, "Failed to create Suricata config file", "FilePath", antreaSuricataConfigPath)
+		return
+	}
+	defer f.Close()
+	if _, err = f.WriteString(suricataAntreaConfigData); err != nil {
+		klog.ErrorS(err, "Failed to write Suricata config file", "FilePath", antreaSuricataConfigPath)
+		return
+	}
+
+	// Open the default Suricata config file /etc/suricata/suricata.yaml.
+	f, err = defaultFS.OpenFile(defaultSuricataConfigPath, os.O_APPEND|os.O_WRONLY, 0600)
+	if err != nil {
+		klog.ErrorS(err, "Failed to open default Suricata config file", "FilePath", defaultSuricataConfigPath)
+		return
+	}
+	defer f.Close()
+	// Include the config file /etc/suricata/antrea.yaml for Antrea in the default Suricata config file /etc/suricata/suricata.yaml.
+	if _, err = f.WriteString(fmt.Sprintf("include: %s\n", antreaSuricataConfigPath)); err != nil {
+		klog.ErrorS(err, "Failed to update default Suricata config file", "FilePath", defaultSuricataConfigPath)
+		return
+	}
+
+	r.startSuricataFn()
+
+	// Wait Suricata command socket file to be ready.
```

Comment:
Is this an official way to wait for suricata main process to be ready? If not, it may happen the rule fails to install?

Reply:
I'll verify that.

Reply:
I found that in Suricata starting log
```
15/12/2022 -- 05:05:27 - <Info> - Found an MTU of 1430 for 'antrea-tap1'
15/12/2022 -- 05:05:27 - <Info> - Found an MTU of 1430 for 'antrea-tap0'
15/12/2022 -- 05:05:27 - <Info> - Running in live mode, activating unix socket
15/12/2022 -- 05:05:27 - <Info> - Using unix socket file '/var/run/suricata/suricata-command.socket'
15/12/2022 -- 05:05:27 - <Notice> - all 16 packet processing threads, 4 management threads initialized, engine started.
15/12/2022 -- 05:05:27 - <Info> - Using BPF 'ip or ip6' on iface 'antrea-tap0'
15/12/2022 -- 05:05:27 - <Info> - Using BPF 'ip or ip6' on iface 'antrea-tap0'
15/12/2022 -- 05:05:27 - <Info> - Using BPF 'ip or ip6' on iface 'antrea-tap0'
15/12/2022 -- 05:05:27 - <Info> - Using BPF 'ip or ip6' on iface 'antrea-tap0'
15/12/2022 -- 05:05:27 - <Info> - Using BPF 'ip or ip6' on iface 'antrea-tap0'
15/12/2022 -- 05:05:27 - <Info> - Using BPF 'ip or ip6' on iface 'antrea-tap0'
15/12/2022 -- 05:05:28 - <Info> - Using BPF 'ip or ip6' on iface 'antrea-tap0'
15/12/2022 -- 05:05:28 - <Info> - Using BPF 'ip or ip6' on iface 'antrea-tap0'
15/12/2022 -- 05:05:28 - <Info> - Using BPF 'ip or ip6' on iface 'antrea-tap1'
15/12/2022 -- 05:05:28 - <Info> - Using BPF 'ip or ip6' on iface 'antrea-tap1'
15/12/2022 -- 05:05:28 - <Info> - Using BPF 'ip or ip6' on iface 'antrea-tap1'
15/12/2022 -- 05:05:28 - <Info> - Using BPF 'ip or ip6' on iface 'antrea-tap1'
15/12/2022 -- 05:05:28 - <Info> - Using BPF 'ip or ip6' on iface 'antrea-tap1'
15/12/2022 -- 05:05:28 - <Info> - Using BPF 'ip or ip6' on iface 'antrea-tap1'
15/12/2022 -- 05:05:28 - <Info> - Using BPF 'ip or ip6' on iface 'antrea-tap1'
15/12/2022 -- 05:05:28 - <Info> - Using BPF 'ip or ip6' on iface 'antrea-tap1'
```
From the log `all 16 packet processing threads, 4 management threads initialized, engine started`, I think this can be used to verify that if the Suricata main process is read.